File: src/main/java/io/vproxy/vfx/ui/stage/CloseButton.java
package io.vproxy.vfx.ui.stage;

import io.vproxy.vfx.theme.Theme;
import javafx.scene.image.Image;
import javafx.scene.layout.CornerRadii;

public class CloseButton extends WindowControlButton {
    private CornerRadii cornerRadii;

    public CloseButton(VStage stage, VStageInitParams initParams) {
        super(stage, initParams);
    }

    @Override
    protected void init(VStageInitParams initParams) {
        if (initParams.iconifyButton || initParams.maximizeAndResetButton) {
            cornerRadii = CornerRadii.EMPTY;
        } else {
            cornerRadii = new CornerRadii(0, 0, 0, 4, false);
        }
    }

    @Override
    protected void onMouseClicked() {
        stage.close();
    }

    @Override
    protected CornerRadii getCornerRadii() {
        return cornerRadii;
    }

    @Override
    protected Image getNormalImage() {
        return Theme.current().windowCloseButtonNormalImage();
    }

    @Override
    protected Image getHoverImage() {
        return Theme.current().windowCloseButtonHoverImage();
    }
}

File: src-ts/src/ui/stage/CloseButton.ts
import { Theme } from '../../theme/Theme.js';
import type { FXImage } from '../../javafx/ImageView.js';
import { WindowControlButton } from './WindowControlButton.js';
import { VStage } from './VStage.js';
import { VStageInitParams } from './VStageInitParams.js';

export class CloseButton extends WindowControlButton {
  private cornerRadiiCssStr = '';

  constructor(stage: VStage, initParams: VStageInitParams) {
    super(stage, initParams);
  }

  protected init(initParams: VStageInitParams): void {
    if (initParams.iconifyButton || initParams.maximizeAndResetButton) {
      this.cornerRadiiCssStr = '0';
    } else {
      this.cornerRadiiCssStr = '0 0 0 4px';
    }
  }

  protected onMouseClicked(): void {
    void this.stage.close();
  }

  protected getCornerRadiiCss(): string { return this.cornerRadiiCssStr; }

  protected getNormalImage(): FXImage {
    return Theme.current().windowCloseButtonNormalImage();
  }
  protected getHoverImage(): FXImage {
    return Theme.current().windowCloseButtonHoverImage();
  }
}
